component/newstorage/mem/mem.go

```diff
@@ -0,0 +1,372 @@
+/*
+Copyright SecureKey Technologies Inc. All Rights Reserved.
+
+SPDX-License-Identifier: Apache-2.0
+*/
+
+package mem
+
+import (
+	"errors"
+	"strings"
+	"sync"
+
+	"github.com/hyperledger/aries-framework-go/pkg/newstorage"
+)
+
+const (
+	expressionTagNameOnlyLength     = 1
+	expressionTagNameAndValueLength = 2
+)
+
+var (
+	errEmptyKey                     = errors.New("key cannot be empty")
+	errInvalidQueryExpressionFormat = errors.New("invalid expression format. " +
+		"it must be in the following format: TagName:TagValue")
+	errIteratorExhausted = errors.New("iterator is exhausted")
+)
+
+// Provider represents an in-memory implementation of the newstorage.Provider interface.
+type Provider struct {
+	dbs map[string]*memStore
+	sync.RWMutex
+}
+
+type closer func(storeName string)
+
+// NewProvider instantiates a new in-memory storage Provider.
+func NewProvider() *Provider {
+	return &Provider{dbs: make(map[string]*memStore)}
+}
+
+// OpenStore opens a store with the given name and returns a handle.
+// If the store has never been opened before, then it is created.
+func (p *Provider) OpenStore(name string) (newstorage.Store, error) {
+	storeName := strings.ToLower(name)
+
+	p.Lock()
+	defer p.Unlock()
+
+	store := p.dbs[storeName]
+	if store == nil {
+		newStore := &memStore{name: storeName, db: make(map[string]dbEntry), close: p.removeStore}
+		p.dbs[storeName] = newStore
+
+		return newStore, nil
+	}
+
+	return store, nil
+}
+
+// SetStoreConfig sets the configuration on a store.
+// The store must be created prior to calling this method.
+func (p *Provider) SetStoreConfig(name string, config newstorage.StoreConfiguration) error {
+	storeName := strings.ToLower(name)
+
+	p.Lock()
+	defer p.Unlock()
+
+	store := p.dbs[storeName]
+	if store == nil {
+		return newstorage.ErrStoreNotFound
+	}
+
+	store.config = config
+
+	return nil
+}
+
+// GetStoreConfig gets the current store configuration.
+// The store must be created prior to calling this method.
+func (p *Provider) GetStoreConfig(name string) (newstorage.StoreConfiguration, error) {
+	storeName := strings.ToLower(name)
+
+	store := p.dbs[storeName]
+	if store == nil {
+		return newstorage.StoreConfiguration{}, newstorage.ErrStoreNotFound
+	}
+
+	return store.config, nil
+}
+
+// Close closes all stores created under this store provider.
+func (p *Provider) Close() error {
+	p.Lock()
+	defer p.Unlock()
+
+	for _, memStore := range p.dbs {
+		memStore.db = make(map[string]dbEntry)
```

Comment:
Should instead call Close on each store.

Reply:
Done